enhancements/local-storage/kubesan-csi-driver-integration-into-lvms.md

```diff
@@ -0,0 +1,452 @@
+  
+3. **Monitoring and Maintenance**:
+  - Continuously monitor the health and status of the VGs.
+  - Handling any required maintenance tasks, such as resizing, repairing, or migrating VGs must be performed manually for shared Volume Groups.
+
+4. **Synchronization**:
+  - Ensure synchronization mechanisms (such as locks) are in place for shared VGs to prevent data corruption and ensure consistency.
+  - Utilize `sanlock` or similar technologies to manage and synchronize access to shared storage at all times.
+  - For SAN lock initialization, a race-free initialization of the lock space will be required. This can be achieved by using a Lease Object,
+    which is a Kubernetes object that can be used to coordinate distributed systems. The Lease Object will be used to ensure that only one node
+    can initialize the lock space at a time. The Lease will be owned on a first-come-first-serve basis, and the node that acquires the Lease will
+    will be used for shared lockspace initialization. [A sample implementation can be found here](https://github.com/openshift/lvm-operator/commit/8ba6307c7bcaccc02953e0e2bdad5528636d5e2d)
+
+
+#### Design Details for Status Reporting
+
+The status reporting will include:
+
+1. **VG Status**:
+  - Report the health and state of each VG managed by `vgmanager`.
+  - Include details such as size, available capacity, and any errors or warnings.
+  - Health reporting per node is still mandatory.
+
+2. **Node-Specific Information**:
+  - Report node-specific information related to the VGs, such as which nodes have access to shared VGs.
+  - Include status of node-local VGs and any issues detected.
+
+3. **CSI Driver Status**:
+  - Provide status updates on the CSI drivers (both TopoLVM and KubeSAN) deployed in the cluster.
+  - Include information on driver health, performance metrics, and any incidents.
+  - Ideally, kubesan implements Volume Health Monitoring CSI calls.
+
+4. **Event Logging**:
+  - Maintain detailed logs of all events related to VG management and CSI driver operations.
+  - Ensure that any significant events (such as failovers, recoveries, and maintenance actions) are logged and reported.
+
+
+### Drawbacks
+
+- Increased complexity in managing both node-local and shared storage.
+- Potential for increased maintenance burden with the integration of a new CSI driver.
+- Risks associated with the stability and maturity of the KubeSAN project.
+- Complex testing matrix and shared volume group use cases can be hard to debug / troubleshoot.
+
+### Topology Considerations
+
+* The primary use case for KubeSAN is to enable shared storage across multiple nodes. This capability is critical for environments where high availability and data redundancy are required.
+* Ensure that all nodes in the cluster can access the shared storage devices consistently and reliably. This may involve configuring network settings and storage paths appropriately.
+
+#### Hypershift / Hosted Control Planes
+
+N/A
```

Comment:
I think that this technology might be useful for Hypershift. Aside from the additional complexity involved, is there any reason that it's excluded?

Reply:
I dont think I follow. Hypershift as of now uses LVMS for its etcd, however the only way a shared VG would be useful would be if Hypershift would be deployed on a cluster where etcd would be stored on a LUN, which is kind of counter intuitive to having a distributed database in the first place.

Could you elaborate what you think would be useful for Hypershift?

Reply:
If youre talking about having local storage access on all virtualized nodes, this is true for any ReadWriteMany CSI drivers, however it doesnt really impact the design of Hypershift in any way.

Reply:
> where etcd would be stored on a LUN

Off the top of my head, a SAN might have performance, availability, or cost characteristics which make it more attractive than local storage. Local storage is probably going to win on performance if cost is no object, but a customer already has a high performance SAN, I can see that being an option they might be interested in considering if we made it available.

Reply:
At that point in time, Hypershift becomes yet another cluster for us, so I dont think this will change the design. Of course we can notify them that this is coming but I dont think they need special consideration

Reply:
I guess to narrow this down: What would you expect me to add to this section? Possible benefits for usage in Hypershift?

Reply:
I asked the hypershift team for their opinion on slack, but I haven't received a response. I don't think this enhancement should be blocked on that though.